api/src/prescription/learner-management/infrastructure/serializers/csv/common-csv-learner-parser.js

```diff
@@ -0,0 +1,199 @@
+  BAD_CSV_FORMAT: 'BAD_CSV_FORMAT',
+  HEADER_REQUIRED: 'HEADER_REQUIRED',
+  HEADER_UNKNOWN: 'HEADER_UNKNOWN',
+  FIELD_MIN_LENGTH: 'FIELD_MIN_LENGTH',
+  FIELD_MAX_LENGTH: 'FIELD_MAX_LENGTH',
+  FIELD_LENGTH: 'FIELD_LENGTH',
+  FIELD_DATE_FORMAT: 'FIELD_DATE_FORMAT',
+  FIELD_EMAIL_FORMAT: 'FIELD_EMAIL_FORMAT',
+  FIELD_REQUIRED: 'FIELD_REQUIRED',
+  FIELD_BAD_VALUES: 'FIELD_BAD_VALUES',
+};
+
+const PARSING_OPTIONS = {
+  header: true,
+  skipEmptyLines: 'greedy',
+  transform: (value) => {
+    if (typeof value === 'string') {
+      value = value.trim();
+      return value.length ? value : undefined;
+    }
+    return value;
+  },
+};
+
+class CommonCsvLearerParser {
+  constructor(input, organizationId, config) {
+    this._input = input;
+    this._organizationId = organizationId;
+    this._errors = [];
+
+    // compute heading
+    this._columns = config.headers;
+
+    // compute validation to handle error
+    this._supportedErrors = config.supportedErrors;
+
+    // compute support_enconding
+    this._supportedEncoding = config.acceptedEncoding;
+  }
+
+  parse(encoding) {
+    if (!encoding) {
+      this._errors.push(new CsvImportError(ERRORS.ENCODING_NOT_SUPPORTED));
+    }
+
+    this.throwHasErrors();
+
+    const { fields } = this._parse(encoding);
+
+    this.throwHasErrors();
+
+    this._checkColumns(fields);
+
+    this.throwHasErrors();
+  }
+
+  throwHasErrors() {
+    if (this._errors.length > 0) throw new AggregateImportError(this._errors);
+  }
+
+  /**
+   * Identify which encoding has the given file.
+   * To check it, we decode and parse the first line of the file with supported encodings.
+   * If there is one with at least "First name" or "Student number" correctly parsed and decoded.
+   */
+  getFileEncoding() {
+    const supported_encodings = ['utf-8', 'win1252', 'macintosh'];
+    const checkedColumns = this._getEncodingColumns();
+    for (const encoding of supported_encodings) {
+      const decodedInput = iconv.decode(this._input, encoding);
+      const {
+        meta: { fields },
+      } = papa.parse(decodedInput, { ...PARSING_OPTIONS, preview: 1 });
+      if (fields.some((value) => checkedColumns.includes(value))) {
+        return encoding;
+      }
+    }
+  }
+
+  _getEncodingColumns() {
+    const checkedColumns = this._columns.filter((c) => c.checkEncoding).map((c) => c.name);
+    if (checkedColumns.length === 0) {
+      return this._columns.map((c) => c.name);
+    }
+    return checkedColumns;
+  }
+
+  _parse(encoding = 'utf8') {
+    const decodedInput = iconv.decode(this._input, encoding);
+    const {
+      data: learnerLines,
+      meta: { fields },
+      errors,
+    } = papa.parse(decodedInput, PARSING_OPTIONS);
+
+    if (errors.length) {
+      const hasDelimiterError = errors.some((error) => error.type === 'Delimiter');
+      if (hasDelimiterError) {
+        this._errors.push(new CsvImportError(ERRORS.BAD_CSV_FORMAT));
+      }
```

Comment:
**question**: Est ce qu'il y a d'autres erreurs qui ne sont pas des `Delimiter` qu'on devrait gérer ?

Reply:
🤷‍♀️ we don't know, on a pris la même règle des gestion que sur les autres fichiers

Reply:
**todo**: prendre en compte les erreurs papaparse ( "Quotes" et "FieldMismatch") et rajouter les cas de test (
```
const lines = `
nom;prenom
lionel

const lines2 = `
nom;prenom
lionel;bé;toto;
```

```
https://www.papaparse.com/docs#errors

Reply:
On a rajouté l'erreur `FieldMismatch` par contre on a pas rajouter l'erreur `Quotes` car 
- cas de test complexe à reproduire
- quand on regarde les autres csv parser, on ne prend pas en compte cette erreur